marker_rustc_driver/src/main.rs

```diff
@@ -49,12 +52,20 @@ struct MarkerCallback;
 
 impl rustc_driver::Callbacks for MarkerCallback {
     fn config(&mut self, config: &mut rustc_interface::Config) {
-        // Clippy explicitly calls any previous functions. This will not be
-        // done here to keep it simple and to ensure that only known code is
-        // executed.
+        // Clippy explicitly calls any previous `register_lints` functions. This
+        // will not be done here to keep it simple and to ensure that only known
+        // code is executed.
         assert!(config.register_lints.is_none());
 
         config.register_lints = Some(Box::new(|_sess, lint_store| {
+            // Register lints from lint crates. This is required to have rustc track
```

Comment:
Are these `thread_local` required as a cache because the callbacks are invoked on per-crate basis, but the values they return for each crate are the same?

Maybe the adapter / converter / lints map, and other stuff could be a field on `MarkerLintPass`? This would avoid some of the thread locals

Reply:
In Marker caches are often used to prevent translating existing notes again and leaking memory. This is a special case, though. Rustc assumes that all lints are static pointers. It uses the memory address as an ID and also for hashing etc. Marker is therefore to keep the memory addresses of these lints static.

The `MarkerLintPass` of the rustc driver is only conditionally constructed later. I tried around with storing it in there before, but then that instance also needed to be stored somewhere. This was the nicest solution I could come up with, but there are most likely better ones.

If you mean the `MarkerLintPass` of the adapter, that's a no, since the rustc lint instance is a concept of the rustc driver.

(The lint passes should probably have different names, now that I searched for it xD
)